daemon/docker_watcher.go

```diff
@@ -258,35 +258,30 @@ func (d *Daemon) handleCreateContainer(id string, retry bool) {
 			continue
 		}
 
-		var orchLabelsModified bool
 		d.containersMU.RLock()
 		cont, ok := d.containers[id]
 		d.containersMU.RUnlock()
-		if ok {
-			cont.Mutex.Lock()
-			if orchLabelsModified = updateOrchLabels(cont, lbls); !orchLabelsModified {
-				log.Debugf("No changes to orch labels, ignoring")
-			}
-		} else {
-			cont = container.NewContainer(dockerContainer, lbls)
-			cont.Mutex.Lock()
+		if !ok {
+			cont = container.NewContainer(dockerContainer)
 		}
 
+		ep.Mutex.Lock()
+		orchLabelsModified := updateOrchLabels(ep, lbls)
+		if ok && !orchLabelsModified {
+			log.Debugf("No changes to orch labels.")
+			ep.Mutex.Unlock()
+			return
+		}
 		// It's mandatory to update the container in its label otherwise
 		// the label will be considered unused.
-		identity, newHash, err := d.updateContainerIdentity(cont.ID, cont.LabelsHash, &cont.OpLabels)
+		identity, newHash, err := d.updateEndpointIdentity(ep.StringID(), ep.LabelsHash, &ep.OpLabels)
 		if err != nil {
 			cont.Mutex.Unlock()
 			log.Warningf("unable to update identity of container %s: %s", id, err)
 			return
 		}
-		cont.LabelsHash = newHash
-		cID := cont.ID
-		cont.Mutex.Unlock()
-
-		if ok && !orchLabelsModified {
```

Comment:
I was hoping you could help me out with this. If you look at the original code, it executes this if statement only if the container already exists. I have removed it initially thinking exactly the same but after many changes I revert to the original logic.